zkevm-circuits/src/state_circuit/constraint_builder.rs

```diff
@@ -226,10 +231,80 @@ impl<F: Field> ConstraintBuilder<F> {
         // TODO: Missing constraints
     }
 
+    fn build_tx_log_constraints(&mut self, q: &Queries<F>) {
+        self.require_equal(
+            "is_write is always true for TxLog",
+            q.is_write.clone(),
+            1.expr(),
+        );
+        self.require_in_set(
+            "field_tag in TxLogFieldTag range",
+            q.field_tag(),
+            set::<F, TxLogFieldTag>(),
+        );
+        // reset log_id when tx_id increases
+        let no_tag_change = 1.expr() - q.tag_change();
+        self.condition(q.id_change() * no_tag_change, |cb| {
+            cb.require_zero(
+                "reset log_id to one when tx_id increases",
+                q.tx_log_id(false) - 1.expr(),
+            );
+            // constrain first field_tag is Address when tx id increases
+            cb.require_equal(
+                "first field_tag is Address",
+                q.field_tag_matches(TxLogFieldTag::Address),
+                1.expr(),
+            );
+            // tx_id increases by 1
+            cb.require_equal("tx_id increases by 1", q.id_change(), 1.expr());
+        });
+
+        self.condition(q.field_tag_matches(TxLogFieldTag::Address), |cb| {
+            cb.require_zero("index is zero for address ", q.tx_log_index(false))
+        });
+
+        // increase log_id when tag changes to Address if tx id stay same
+        self.condition(
+            q.is_id_unchanged.clone()
+                * q.tag_matches(RwTableTag::TxLog, true)
```

Comment:
we can add a col `is_tag_unchanged` to avoid use lagrange twice bringing degree to 28...

Reply:
updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/552/commits/3017da9d864edd3d8e5b47fee97f34c94bb1f5e6 thanks !!